submissions/daniil-bodryagin/oop-exercise/index-oop.js

```diff
@@ -0,0 +1,180 @@
+import { print } from './js/lib.js';
+
+class Creature {
+  constructor ({name, gender, legs}) {
+    this.name = name;
+    this.gender = gender;
+    this.legs = legs;
+    this._friends = [];
+  }
+
+  set friends(listOfFriends) {
+    this._friends = listOfFriends;
+  }
+
+  get friends() {
+    return this._friends.map(({name}) => name).join(', ');
+  }
+
+  toString() {
+    const {constructor: {name: species}, name, gender, legs, friends} = this;
+    return [species, name, gender, legs, friends].join('; ');
+  }
+}
+
+class Human extends Creature {
+  constructor ({name, gender, legs = 2, hands = 2, phrase = 'Hello!'}) {
+    super({name, gender, legs});
+    this.phrase = phrase;
+    this.hands = hands;
+  }
+
+  toString() {
+    const {hands, phrase} = this;
+    const creatureString = super.toString();
+    const [species, name, gender, legs, friends] = creatureString.split('; ');
+    return [species, name, gender, legs, hands, phrase, friends].join('; ');
+  }
+}
+
+class Cat extends Creature {
+  constructor ({name, gender, legs = 4}) {
+    super({name, gender, legs});
+    canMeow(this);
+  }
+
+  toString() {
+    const {phrase} = this;
```

Comment:
`phrase` is obviously a shared property just as legs and or name or friends.
Why treated differently?

Reply:
In the `Cat` and `Dog` classes `phrase` is the only property which is absent in base class. In `Human` class they are `hands` and `phrase` respectively.
Why didn't I put this property in the base class? Because every cat must say the same - "Meow" (as it was written in the assignment), and if it would be part of base constructor I would be able to pass different phrases as arguments every time I create new class instance.
Why didn't I put this property in child classes with predefined value? I wanted to do this, but the obstacle is `Catwoman` class. I needed to borrow `phrase` property from `Cat` that's why I created special functions just as canMeow() which sets the phrase value to cats and catwomen. Since humans can say different phrases they have this property in constructor.
Why didn't I use function to implement composition only for catwoman and didn't leave cats and dogs with `phrase` property in costructors? It seemed to me that code would look inconsistent then and that it would be right to use special function for adding phrase property everywhere.

Reply:
> Why didn't I put this property in the base class? Because every cat must say the same - "Meow" (as it was written in the assignment), and if it would be part of base constructor I would be able to pass different phrases as arguments every time I create new class instance.

This is solvable by sending a relevant value to the parent's constructor.

> Why didn't I put this property in child classes with predefined value? I wanted to do this, but the obstacle is Catwoman class. I needed to borrow phrase property from Cat that's why I created special functions just as canMeow() which sets the phrase value to cats and catwomen. Since humans can say different phrases they have this property in constructor.

... and that can be `canMeow()` (probably a bit refactored), i.e. `super(.... canMeow()...)` is a valid approach.
Also you may want to look at static class methods to solve the problem with CatWoman's phrase. `Cat.getPhrase()` is even more expressive than `canMeow` from the global context.

`canMeow` is the right approach for compositional pattern (as opposed to hierarchical approach of OOP). That's exactly the catch here.

Reply:
Added `phrase` to base class, replaced global functions by static class methods. Now the code looks better indeed. Thanks!